enhancements/insights/on-demand-data-gathering.md

```diff
@@ -0,0 +1,255 @@
+	// If all the gatherers are disabled, the Insights operator does not gather any data.
+	// The particular gatherers IDs can be found at https://github.com/openshift/insights-operator/blob/master/docs/gathered-data.md.
+	// Run the following command to get the names of last active gatherers:
+	// "oc get insightsoperators.operator.openshift.io cluster -o json | jq '.status.gatherStatus.gatherers[].name'"
+	// An example of disabling gatherers looks like this: `disabledGatherers: ["clusterconfig/machine_configs", "workloads/workload_info"]`
+	// +optional
+	DisabledGatherers []string `json:"disabledGatherers"`
+}
+
+const (
+	// No data obfuscation
+	NoPolicy DataPolicy = "None"
+	// IP addresses and cluster domain name are obfuscated
+	ObfuscateNetworking DataPolicy = "ObfuscateNetworking"
+	// Data gathering is running
+	Running State = "Running"
+	// Data gathering is completed
+	Completed State = "Completed"
+	// Data gathering failed
+	Failed State = "Failed"
+)
+
+// dataPolicy declares valid data policy types
+// +kubebuilder:validation:Enum="";None;ObfuscateNetworking
+type DataPolicy string
+
+// state declares valid gathering state types
+// +kubebuilder:validation:Enum=Running;Completed;Failed
+type State string
+
+type DataGatherStatus struct {
+	// state reflects the current state of the data gathering process.
+	// +kubebuilder:validation:Required
+	State State `json:"state"`
+	// gatherers is a list of active gatherers (and their statuses) in the last gathering.
+	// +listType=atomic
+	// +kubebuilder:validation:Required
+	Gatherers []GathererStatus `json:"gatherers"`
+	// startTime is the time when Insights data gathering started.
+	// +kubebuilder:validation:Required
+	StartTime metav1.Time `json:"startTime"`
+	// finishTime is the time when Insights data gathering finished.
+	// +kubebuilder:validation:Required
+	FinishTime metav1.Time `json:"finishTime"`
+}
+
+// gathererStatus represents information about a particular
+// data gatherer.
+type GathererStatus struct {
+	// conditions provide details on the status of each gatherer.
+	// +listType=atomic
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinItems=1
+	Conditions []metav1.Condition `json:"conditions"`
+	// name is the name of the gatherer.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MaxLength=256
+	// +kubebuilder:validation:MinLength=5
+	Name string `json:"name"`
+	// lastGatherDuration represents the time spent gathering.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Type=string
+	// +kubebuilder:validation:Pattern="^([1-9][0-9]*(\\.[0-9]+)?(ns|us|µs|ms|s|m|h))+$"
+	LastGatherDuration metav1.Duration `json:"lastGatherDuration"`
+}
+
+```
+
+### Workflow Description
+
+1. User/cluster admin creates a new CR instance of the `datagather.insights.openshift.io` CRD. 
+2. The Insights operator is informed (using informers) about the new CR instance and spawns a new data gathering in a new container. 
```

Comment:
Yes this would allow parallel running. I don't see much benefit in that though, because (as described below) the processing part is not designed for multiple results for one cluster - i.e there's no history of the Insights results (in cluster) and I think it doesn't make much sense. I think user is interested in the latest available Insights result.

Reply:
> I think user is interested in the latest available Insights result.

Given the ability to select a subset or a different set of gatherers, I think it's reasonable to want both the latest and a previously created with different gatherers.  In light of that, some parallelism, with priority for the latest seems reasonable.

Reply:
> processing part is not designed for multiple results for one cluster

I feel like we need to have a conversation with the owners of this part to see if that can be unblocked, as it makes the UX here (and the implementation I think) much more difficult

Reply:
IMO it's a conversation worth having, perhaps one of the subteam could comment here to explain why the current model is 1 per cluster and what is the level of effort to make it more flexible. Having a record of that and using that as justification for the queue approach within the enhancement would be helpful for future readers IMO

Reply:
@JoelSpeed it is done that way because the requirements is to show most recent recommendation in HAC (previously in OCM) and in ACM. There was not any need to remember the history of recommendations. On other hand we also operate CCX Notification Service that needs to remember older results with all the consequences (much higher DB usage, some performance issues etc.). Is there any requirement to store the (whole?) history of recommendations per cluster in external (customer-visible) service(s)?

Reply:
> Is there any requirement to store the (whole?) history of recommendations per cluster in external (customer-visible) service(s)?

This new feature has been requested so that users can trigger on demand insights gathering. The feature as it stands currently is proposed to allow users to capture these and then remove them after 24 hours (I think that's what we agreed).

Given that the product already periodically gathers the data every few hours anyway, this means that should a user create the on-demand gather, their data is available for at most the period between data gathers in the current model.

I would expect this feature to be much more useful if the data was available for some minimum period, if we agree that's 24/28/72 hours (some sensible cap), I think the UX of this feature is greatly improved.

The alternatives seem to be that either, we queue the jobs and make sure they never run in parallel (delaying any scheduled/periodic jobs), or we cancel any that are in progress when a new one is created, which also seems poor.

@tremes When a customer requests an on-demand gather, do we expect them to ever want to look directly at the console.redhat.com result or do we expect them to only look at the CR? If they want to look at the console link explicitly, I think it's a major UX win if that could be persisted for the lifetime of the CR

Reply:
My feeling is that this whole work (as well as the previous with the new Insights CRDs) is to provide more information about the data gathering in cluster, but yes I think we expect customers/users to consume the content from console.redhat.com (simply because this is the only place with the full/complete recommendation content), but I still can't see any major advantage in providing some "Insights history" in the console.redhat.com. My take is that if you're interested (for some reason) in the previous Insights results then go to the cluster and check the corresponding CR. This is IMHO a good first step (at least within the TechPreview feature) and then we can discuss the "evolution".

Reply:
I can see the motivation in keeping it simple for tech preview, but I do feel strongly about the UX here.

If I as a user create something, that then creates some artifacts, I should be able to recover those for the duration of my resource existing. What would suck UX wise is if I create a gather, and want to go look at my console result once it's done, after 15 minutes I come back to check on my results (I went to get a coffee?) and then I see my results have been pruned already, because someone else ran their own insights gather, frustrating! Next I'm going to have to re-run mine and hope they don't re-run theirs in the mean time as well.

The only way to solve that seems to be to have concurrent access to history for at least some short period

Reply:
Where did this close?  The `InsightsRequestID` seems to make parallel possible even if not required. (I like the inclusion of that field)

Reply:
The parallelism is not forbidden (it's jut not explicitly mentioned that it's allowed. Do you think we should mention it?).  I removed the wording about queueing here https://github.com/openshift/enhancements/pull/1279/commits/7c30fc88ed49814e2bff0baee7d1ec6d133a9760#diff-bbc5c23b6f7f1f49d1b7832ef1213925909b286f76fea0d67c81cafebfa28049R242

Reply:
Well I see the queuing is mentioned on more places so let me update it.

Reply:
This has been updated. Queu is not mentioned. Jobs can be parallel, but there's this `InsightsRequestID` allowing to check the state of the Insights analysis.